xcvm/lib/core/src/asset.rs

```diff
@@ -0,0 +1,214 @@
+
+impl From<AssetID> for u32 {
+	fn from(val: AssetID) -> Self {
+		val.0
+	}
+}
+
+impl From<u32> for AssetID {
+	fn from(asset: u32) -> Self {
+		AssetID(asset)
+	}
+}
+
+impl From<PICA> for AssetID {
+    fn from(_: PICA) -> Self {
+      PICA::ID
+    }
+}
+
+impl From<ETH> for AssetID  {
+    fn from(_: ETH) -> Self {
+      ETH::ID
+    }
+}
+
+impl From<USDT> for AssetID {
+    fn from(_: USDT) -> Self {
+      USDT::ID
+    }
+}
+
+impl From<USDC> for AssetID {
+    fn from(_: USDC) -> Self {
+      USDC::ID
+    }
+}
+
+pub struct InvalidAsset;
+pub struct PICA;
+pub struct ETH;
+pub struct USDT;
+pub struct USDC;
+
+/// List of XCVM compatible networks.
+/// The order matter and must not be changed, adding a network on the right is safe.
+pub type Assets = (InvalidAsset, (PICA, (ETH, (USDT, (USDC, ())))));
+
+/// Type implement network must be part of [`Networks`], otherwise invalid.
+pub trait Asset {
+	const ID: AssetID;
+}
+
+impl Asset for PICA {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for ETH {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for USDT {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for USDC {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Encode, Decode, TypeInfo, Serialize, Deserialize,
+)]
+#[repr(transparent)]
+pub struct Displayed<T>(
+	#[serde(bound(serialize = "T: core::fmt::Display"))]
+	#[serde(serialize_with = "serialize_as_string")]
+	#[serde(bound(deserialize = "T: core::str::FromStr"))]
+	#[serde(deserialize_with = "deserialize_from_string")]
+	pub T,
+);
+
+fn serialize_as_string<S: Serializer, T: core::fmt::Display>(
+	t: &T,
+	serializer: S,
+) -> Result<S::Ok, S::Error> {
+	serializer.serialize_str(&t.to_string())
+}
+
+fn deserialize_from_string<'de, D: Deserializer<'de>, T: core::str::FromStr>(
+	deserializer: D,
+) -> Result<T, D::Error> {
+	let s = alloc::string::String::deserialize(deserializer)?;
+	s.parse::<T>().map_err(|_| serde::de::Error::custom("Parse from string failed"))
+}
+
+impl<T> From<T> for Displayed<T> {
+	fn from(x: T) -> Self {
+		Displayed(x)
+	}
+}
+
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Encode, Decode, TypeInfo, Serialize, Deserialize,
+)]
+#[serde(rename_all = "snake_case")]
+pub enum Amount {
+	Fixed(Displayed<u128>),
+	Ratio(u32),
+}
+
+impl From<u128> for Amount {
+	fn from(x: u128) -> Self {
+		Amount::Fixed(Displayed(x))
+	}
+}
+
+impl Amount {
+	pub fn apply(&self, value: u128) -> u128 {
+		match self {
+			Amount::Fixed(Displayed(x)) => *x,
+			Amount::Ratio(x) => FixedU128::<U16>::from_num(value)
+				.saturating_mul(FixedU128::<U16>::from_num(*x as u128).saturating_div(FixedU128::<
+					U16,
```

Comment:
just it get this, ratio only can represend 1/100? i mean if i seen 1000, I can tell 990 or 10, not 999 or 1? Either fee or expected return?

Reply:
Oh, it's legacy, must be u32::MAX

Reply:
Nice catch